### PR TITLE
Prepare MultivariateGaussianConstraint to accept a response matrix for folded data

### LINKED ISSUE #175: Add a ForwardFoldedGaussianConstraint
Relevant e.g. for the recent Belle analysis on B->D^*lnu.

### PR DESCRIPTION
**Pull request by dvandyk**
Modifications produced together with Chrístoph Bobeth.

GitHub: resolves #175

### CODE REVIEW

File: eos/statistics/log-likelihood.cc [hunk 1]
```diff
@@ -1021,7 +1024,10 @@ namespace eos
                 gsl_matrix * covariance = gsl_matrix_alloc(k, k);
                 gsl_matrix_memcpy(covariance, _covariance);
 
-                return LogLikelihoodBlockPtr(new MultivariateGaussianBlock(cache, std::move(ids), mean, covariance, _number_of_observations));
+                gsl_matrix * response = gsl_matrix_alloc(k, k);
```
Comment (cBobeth) [inline eos/statistics/log-likelihood.cc]: see comment L818

File: eos/statistics/log-likelihood.hh [hunk 2]
```diff
@@ -250,7 +250,11 @@ namespace eos
                     }
                 }
 
-                return LogLikelihoodBlock::MultivariateGaussian(cache, _observables, _mean, _covariance, number_of_observations);
+                // create GSL matrix for the response matrix
+                gsl_matrix * _response = gsl_matrix_calloc(n_, n_);
```
Comment (cBobeth) [inline eos/statistics/log-likelihood.hh]: see previous comment

File: eos/statistics/log-likelihood.hh [hunk 3]
```diff
@@ -289,7 +293,11 @@ namespace eos
                     }
                 }
 
-                return LogLikelihoodBlock::MultivariateGaussian(cache, _observables, _mean, covariance, number_of_observations);
+                // create GSL matrix for the response matrix
+                gsl_matrix * _response = gsl_matrix_calloc(n_, n_);
```
Comment (cBobeth) [inline eos/statistics/log-likelihood.hh]: see previous comment

File: eos/constraint.cc [hunk 4]
```diff
@@ -1081,28 +1081,29 @@ namespace eos
             means(means),
             covariance(covariance),
             number_of_observations(number_of_observations),
-            dim(observables.size())
+            dim_meas(means.size()),
+            dim_pred(observables.size())
         {
-            if (dim != kinematics.size()) { throw InternalError("MultivariateGaussianConstraintEntry: wrong number of kinematics"); }
+            if (dim_meas != dim_pred) { throw InternalError("MultviateGaussianConstraintEntry: number of measurements does not equal number of predictions"); }
```
Comment (cBobeth) [inline eos/constraint.cc]: Multiviate -> Multivariate

File: eos/constraint.cc [hunk 5]
```diff
@@ -1081,28 +1081,29 @@ namespace eos
...
 
-            if (dim != options.size()) { throw InternalError("MultivariateGaussianConstraintEntry: wrong number of options"); }
+            if (dim_meas != covariance.size()) { throw InternalError("MultviateGaussianConstraintEntry: number of rows in covariance does not equal number of measurements"); }
```
Comment (cBobeth) [inline eos/constraint.cc]: see L1087

File: eos/constraint.cc [hunk 6]
```diff
@@ -1081,28 +1081,29 @@ namespace eos
...
+            for (auto i = 0u ; i < dim_meas ; ++i)
+            {
+                if (dim_meas != covariance[i].size()) { throw InternalError("MultviateGaussianConstraintEntry: number of columns in covariance row " + stringify(i) + " does not equal number of measurements"); }
```
Comment (cBobeth) [inline eos/constraint.cc]: see L1087

File: eos/constraint.cc [hunk 7]
```diff
@@ -1072,25 +1074,27 @@ namespace eos
...
             options(options),
             means(means),
             covariance(covariance),
+            response(response),
             number_of_observations(number_of_observations),
             dim_meas(means->size),
             dim_pred(observables.size())
         {
-            if (dim_meas != dim_pred) { throw InternalError("MultviateGaussianConstraintEntry: number of measurements does not equal number of predictions"); }
+            if ((nullptr == response) && (dim_meas != dim_pred)) { throw InternalError("MultviateGaussianConstraintEntry: number of measurements does not equal number of predictions in absence of a response matrix"); }
```
Comment (cBobeth) [inline eos/constraint.cc]: Multviate -> Multivariate

File: eos/constraint.cc [hunk 8]
```diff
@@ -1072,25 +1074,27 @@ namespace eos
...
 
-            if (dim_meas != covariance.size()) { throw InternalError("MultviateGaussianConstraintEntry: number of rows in covariance does not equal number of measurements"); }
-            for (auto i = 0u ; i < dim_meas ; ++i)
-            {
-                if (dim_meas != covariance[i].size()) { throw InternalError("MultviateGaussianConstraintEntry: number of columns in covariance row " + stringify(i) + " does not equal number of measurements"); }
-            }
+            if (dim_meas != covariance->size1) { throw InternalError("MultviateGaussianConstraintEntry: number of rows in covariance does not equal number of measurements"); }
+            if (dim_meas != covariance->size2) { throw InternalError("MultviateGaussianConstraintEntry: number of columns in covariance does not equal number of measurements"); }
```
Comment (cBobeth) [inline eos/constraint.cc]: see L1091

File: eos/constraint.cc [hunk 9]
```diff
@@ -1072,25 +1074,27 @@ namespace eos
...
+
+            if ((nullptr != response) && (dim_meas != response->size1)) { throw InternalError("MultviateGaussianConstraintEntry: number of rows in response does not equal number of measurements");}
```
Comment (cBobeth) [inline eos/constraint.cc]: see L1091

File: eos/constraint.cc [hunk 10]
```diff
@@ -815,7 +815,10 @@ namespace eos
                 }
             }
 
-            LogLikelihoodBlockPtr block = LogLikelihoodBlock::MultivariateGaussian(cache, observables, means, covariance, number_of_observations);
+            gsl_matrix * response = gsl_matrix_calloc(dim, dim);
```
[the remark this answers is not in the document]
Comment (dvandyk) [reply eos/constraint.cc]: That line is in ```MultivariateGaussianConstraintEntry```, not the multivariate case that we modified.

File: eos/constraint.cc [hunk 11]
```diff
@@ -1134,7 +1137,11 @@ namespace eos
                 }
             }
 
-            auto block = LogLikelihoodBlock::MultivariateGaussian(cache, observables, means, covariance, number_of_observations);
+            // create GSL matrix for the response
+            gsl_matrix * response = gsl_matrix_calloc(dim, dim);
```
[the remark this answers is not in the document]
Comment (dvandyk) [reply eos/constraint.cc]: This is the wrong version of the file you looked into. It says "outdated" in the Github interface...

File: eos/statistics/log-likelihood.cc (hunk 1 above)
Comment (dvandyk) [reply eos/statistics/log-likelihood.cc]: This is a bug. Thanks, fixed locally. Will push in a bit.

File: eos/constraint.cc (hunk 4 above)
Comment (dvandyk) [reply eos/constraint.cc]: Fixed!

File: eos/constraint.cc (hunk 5 above)
Comment (dvandyk) [reply eos/constraint.cc]: Fixed

File: eos/constraint.cc (hunk 6 above)
Comment (dvandyk) [reply eos/constraint.cc]: Fixed

File: eos/constraint.cc (hunk 9 above)
Comment (dvandyk) [reply eos/constraint.cc]: Fixed

File: eos/constraint.cc (hunk 7 above)
Comment (dvandyk) [reply eos/constraint.cc]: Fixed

File: eos/constraint.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (dvandyk) [reply eos/constraint.cc]: You looked at one of the intermediary commits I think. Marking as resolved.

**Pull request merged**